File: src/main/java/sample/design_pattern/Composite.java
package sample.design_pattern;

import java.util.*;


abstract class Component {
    abstract public void printAll();
    public List<Component> getComponentList() {
        return new ArrayList<Component>();
    }
}


class Directory extends Component {
    private String name;
    private List<Component> children;

    public Directory(String name, Component... args) {
        this.name = name;

        children = new ArrayList<Component>();
        for (Component component : args) {
            children.add(component);
        }
    }

    public void printAll() {
        System.out.println("Directory name: " + name);

        for (Component component : children) {
            component.printAll();
        }
    }

    public List<Component> getComponentList() {
        List<Component> result = new ArrayList<Component>();

        for (Component component : children) {
            result.addAll(component.getComponentList());
        }

        return result;
    }
}

class File extends Component {
    private String name;

    public File(String name) {
        this.name = name;
    }

    public void printAll() {
        System.out.println("    file name: " + name);
    }
}


class Composite {
    public static void main(String[] args) {
        Component parent = new Directory("User",
            new Directory("hospital", new File("ouch")),
            new File("hello world"),
            new File("sorry")
        );
        parent.printAll();
    }
}

File: src/main/scala/sample/design_pattern/Composite.scala
/*
 * Copyright 2013 SatoShun
 *
 * Composite Pattern
 */

package sample.design_pattern;


trait Component {
    def printAll(): Unit;

    def getComponentList(): List[Component] = {
        List.empty[Component]
    }
}


class Directory(val name: String, val args: Component*) extends Component {
    def printAll() = {
        println("Directory name : " + name)

        args.foreach(_.printAll())
    }

    override def getComponentList(): List[Component] = {
        var result = List.empty[Component]

        args.map({e => result = result ::: e.getComponentList()})
        result
    }
}


class File(val name: String) extends Component {
    def printAll() = println("\tfile name : " + name)

    override def getComponentList(): List[Component] = {
        List(this)
    }
}


object Composite {
    def main(args: Array[String]) = {
        val root = new Directory("User",
            new Directory("hospital", new File("ouch")),
            new File("hello world"),
            new File("sorry")
        )
        root.printAll()

        println(root.getComponentList())
    }
}
